cosima_cookbook/explore.py

```diff
@@ -0,0 +1,965 @@
+        elements
+        """
+        if selector.new is None:
+            return
+        self._show_experiment_information(selector.new)
+
+    def _show_experiment_information(self, experiment_name):
+        """
+        Populate box with experiment information
+        """
+        expt = self.de.experiments[self.de.experiments.experiment == experiment_name]
+
+        style ="""
+        <style>
+            body  { font: normal 8px Verdana, Arial, sans-serif; }
+            table { padding: 10px; border-spacing: 0px 0px; background-color: #fff;" }
+            td    { padding: 10px; }
+            p     { line-height: 1.2; margin-top: 5px }
+        </style>
+        """
+        self.widgets['expt_info'].value = style + """
+        <table>
+        <tr><td><b>Experiment:</b></td> <td>{experiment}</td></tr>
+        <tr><td style="vertical-align:top;"><b>Description:</b></td> <td><p>{description}</p></td></tr>
+        <tr><td><b>Notes:</b></td> <td>{notes}</td></tr>
+        <tr><td><b>Contact:</b></td> <td>{contact} &lt;{email}&gt;</td></tr>
+        <tr><td><b>No. files:</b></td> <td>{nfiles}</td></tr>
+        <tr><td><b>Created:</b></td> <td>{created}</td></tr>
+        </table>
+        """.format(
+                   experiment=experiment_name,
+                   description=return_value_or_empty(expt.description.values[0]),
+                   notes=return_value_or_empty(expt.notes.values[0]),
+                   contact=return_value_or_empty(expt.contact.values[0]),
+                   email=return_value_or_empty(expt.email.values[0]),
+                   nfiles=return_value_or_empty(expt.ncfiles.values[0]),
+                   created=return_value_or_empty(expt.created.values[0]),
+                   )
+        
+    def _filter_experiments(self, b):
+        """
+        Filter experiment list by keywords and variable
+        """
+        options = set(self.de.experiments.experiment)
+
+        kwds = self.widgets['filter_widget'].value
+        if len(kwds) > 0:
+            options.intersection_update(self.de.keyword_filter(kwds))
+
+        variables = self.widgets['var_filter'].selected_vars()
+        if len(variables) > 0:
+            options.intersection_update(self.de.variable_filter(variables))
+
+        self.widgets['expt_selector'].options = sorted(options, key=str.casefold)
+
+    def _load_experiment(self, b):
+        """
+        Open an Experiment Explorer UI with selected experiment
+        """
+        if self.widgets['expt_selector'].value is not None:
+            self.ee = ExperimentExplorer(session=self.session, 
+                                         experiment=self.widgets['expt_selector'].value)
+            self.widgets['expt_explorer'].children = [self.ee]
+
+
+class ExperimentExplorer(VBox):
+
+    session = None
+    data = None
+    experiment_name = None
+    variables = []
+    widgets = {}
+    handlers = {}
+
+    def __init__(self, session=None, experiment=None):
```

Comment:
`self.session` is never set?

Reply:
Yeah I picked that up when I was writing tests. Thanks for spotting though.